connect-inject/health_check_resource.go

```diff
@@ -0,0 +1,311 @@
+		Name:      consulHealthCheckID,
+		Notes:     "Kubernetes Health Check " + reason,
+		ServiceID: serviceID,
+		AgentServiceCheck: api.AgentServiceCheck{
+			TTL:                            "100000h",
+			Status:                         status,
+			Notes:                          reason,
+			TLSSkipVerify:                  true,
+			SuccessBeforePassing:           1,
+			FailuresBeforeCritical:         1,
+			DeregisterCriticalServiceAfter: "",
+		},
+	})
+	if err != nil {
+		h.Log.Error("unable to register health check with Consul from k8s: %v", err)
+		return err
+	}
+	return nil
+}
+
+// getServiceCheck will return the health check for this pod+service or nil if it doesnt exist yet
+func (h *HealthCheckResource) getServiceCheck(client *api.Client, pod *corev1.Pod, serviceID, healthCheckID string) (*api.AgentCheck, error) {
+	filter := fmt.Sprintf("Name == `%s`", healthCheckID)
+	checks, err := client.Agent().ChecksWithFilter(filter)
+	if err != nil {
+		h.Log.Error("unable to get agent health checks for %v, %v", healthCheckID, filter, err)
+		return nil, err
+	}
+	// This will be nil (does not exist) or an actual check!
+	return checks[healthCheckID], nil
+}
+
+// getReadyStatusAndReason returns the formatted status string to pass to Consul based on the
+// ready state of the pod along with the reason message which will be passed into the Notes
+// field of the Consul health check.
+func (h *HealthCheckResource) getReadyStatusAndReason(pod *corev1.Pod) (string, string, error) {
+	for _, cond := range pod.Status.Conditions {
+		var consulStatus, reason string
+		if cond.Type == corev1.PodReady {
+			if cond.Status != corev1.ConditionTrue {
+				consulStatus = healthCheckCritical
+				reason = cond.Message
+			} else {
+				consulStatus = healthCheckPassing
+				reason = kubernetesSuccessReasonMsg
+			}
+			return consulStatus, reason, nil
+		}
+	}
+	return "", "", fmt.Errorf("no ready status for pod: %s", pod.Name)
+}
+
+// getConsulClient returns a new *api.Client pointed at the consul agent local to the pod
+func (h *HealthCheckResource) getConsulClient(pod *corev1.Pod) (*api.Client, error) {
+	httpFmt := "http"
+	if h.ConsulPort == "8501" {
+		httpFmt = "https"
+	}
+	newAddr := fmt.Sprintf("%s://%s:%s", httpFmt, pod.Status.HostIP, h.ConsulPort)
+	localConfig := h.ClientConfig
+	localConfig.Address = newAddr
+	localClient, err := api.NewClient(localConfig)
+	if err != nil {
+		h.Log.Error("unable to get Consul API Client for address %s: %s", newAddr, err)
+		return nil, err
+	}
+	h.Log.Debug("setting consul client to the following agent: %v", newAddr)
+	return localClient, err
+}
+
+// shouldProcess is a simple filter which determines if Upsert should attempt to process the pod
+// this is done without making any client api calls so it is fastpath
+func (h *HealthCheckResource) shouldProcess(pod *corev1.Pod) bool {
+	return pod.Annotations[annotationInject] == "true" && pod.Status.Phase == corev1.PodRunning
```

Comment:
I can switch the annotation no problem.
I guess I wanted a sanity check that it was actually connect-injected and not mistakenly just carrying a pod label around from a yaml file, in which case the pod wont be registered and all the calls will fail and requeue over and over.

Reply:
I think we can safely assert on the `consul.hashicorp.com/injected-status` annotation because that's "owned" by us since it's within the `consul.hashicorp.com` prefix.

Reply:
Cool injected-status it is!

Reply:
Yeah, I think if you want to do an extra check, then you would need to use the status annotation (aka `consul.hashicorp.com/injected-status` or the `annotationStatus` constant) because the `annotationInject` is only present if the user has provided it. However, there are cases when this annotation could be absent, but the pods will still be injected. This can happen when the helm chart is [configured](https://www.consul.io/docs/k8s/helm#v-connectinject-default) to inject all pods by default. In this case, the health check controller will ingore these pods even though it shouldn't.

Reply:
makes sense! I've pushed that change then, thanks for the explanation!